#Atividade prática com datasets de energia — Orange Data Mining, Python e Pandas

###Nome dos Integrantes:

- João Pedro Ferrari Faian        RM:573037
- Vitor Nascimento                RM:571873
- Lucas Santana                   RM:573197
- Pierre Biason                   RM:569718
- Lucca Bracco                    RM:570175
- Nickollas Korner                RM:569655

##Importando as bibliotecas:

In [ ]:
import pandas as pd

## Dataset 1 — Appliances Energy Prediction (UCI)



##### Fonte: https://archive.ics.uci.edu/dataset/374/appliances+energy+prediction

**Situação**

Uma empresa de eficiência energética está analisando o comportamento de uma residência de baixo consumo. A equipe deseja identificar períodos de consumo elevado dos eletrodomésticos e observar quais condições de temperatura e umidade estavam presentes nesses momentos.

In [ ]:
#Realizar a importacao do arquivo
dados = pd.read_csv('exercicio1_sample.csv')

# 1. Apresente head(), shape, info() e describe()
print("--- Primeiras linhas ---")
display(dados.head())
print("\n--- Dimensões (Linhas, Colunas) ---")
print(dados.shape)
print("\n--- Estrutura e Tipos de Dados ---")
dados.info()
print("\n--- Estatísticas Descritivas ---")
display(dados.describe())

--- Primeiras linhas ---


,date,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
0,time,continuous,continuous,continuous,continuous,continuous,continuous,continuous,continuous
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2016-03-14 01:40:00,40,0,20.89,35.4,17.76,39.1633,20.29,36.9
3,2016-01-30 20:00:00,90,10,21.89,53.1,21.29,45.36,21.6333,49.2267
4,2016-03-15 03:00:00,50,0,21.39,35.5,17.6333,40.53,21.6667,35.2



--- Dimensões (Linhas, Colunas) ---
(13817, 9)

--- Estrutura e Tipos de Dados ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13817 entries, 0 to 13816
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        13816 non-null  object
 1   Appliances  13816 non-null  object
 2   lights      13816 non-null  object
 3   T1          13816 non-null  object
 4   RH_1        13816 non-null  object
 5   T2          13816 non-null  object
 6   RH_2        13816 non-null  object
 7   T3          13816 non-null  object
 8   RH_3        13816 non-null  object
dtypes: object(9)
memory usage: 971.6+ KB

--- Estatísticas Descritivas ---


,date,Appliances,lights,T1,RH_1,T2,RH_2,T3,RH_3
count,13816,13816,13816,13816,13816,13816,13816,13816,13816
unique,13816,90,8,605,2188,1344,2745,1147,2156
top,2016-02-03 14:20:00,50,0,21,39.09,19.2,40,22.39,37.59
freq,1,3054,10686,370,108,250,90,319,170


In [ ]:
# 2. Renomeie Appliances para Consumo_Eletrodomesticos e simplifique atributos ambientais
dados = dados.rename(columns={
    "Appliances": "Consumo_Eletrodomesticos",
    "T1": "Temp_Cozinha",
    "RH_1": "Umid_Cozinha",
    "T2": "Temp_Sala",
    "RH_2": "Umid_Sala",
    "T3": "Temp_Lavabo",
    "RH_3": "Umid_Lavabo",
})

print(dados.columns)

Index(['date', 'Consumo_Eletrodomesticos', 'lights', 'Temp_Cozinha',
       'Umid_Cozinha', 'Temp_Sala', 'Umid_Sala', 'Temp_Lavabo', 'Umid_Lavabo'],
      dtype='object')


In [ ]:
# 3. Determine o maior consumo de eletrodomésticos registrado na amostra
consumo_maximo = dados['Consumo_Eletrodomesticos'].max()
print(f"Maior consumo registrado: {consumo_maximo} Wh")

# 4. Calcule um limiar correspondente a 70% do valor máximo
limiar_70 = 0.70 * consumo_maximo
print(f"Limiar de 70% do máximo: {limiar_70} Wh")
df_alto_consumo = dados[dados['Consumo_Eletrodomesticos'] > limiar_70]

# 5. Conte os registros selecionados e calcule qual percentual da amostra eles representam
qtd_alto_consumo = len(df_alto_consumo)
percentual_alto = (qtd_alto_consumo / len(dados)) * 100

print(f"Quantidade de registros acima de 70%: {qtd_alto_consumo}")
print(f"Percentual que representam da amostra: {percentual_alto:.2f}%")

Maior consumo registrado: 1080 Wh
Limiar de 70% do máximo: 756.0 Wh
Quantidade de registros acima de 70%: 22
Percentual que representam da amostra: 0.16%


In [ ]:
# 6. Calcule a temperatura média de T1 (que renomeamos para Temp_Cozinha)
temp_media = dados['Temp_Cozinha'].mean()
print(f"Temperatura média (T1): {temp_media:.2f}°C")

df_consumo_e_temp = dados[
    (dados['Consumo_Eletrodomesticos'] > limiar_70) &
    (dados['Temp_Cozinha'] > temp_media)
]

qtd_consumo_e_temp = len(df_consumo_e_temp)
print(f"Registros com alto consumo E temperatura acima da média: {qtd_consumo_e_temp}")

Temperatura média (T1): 21.69°C
Registros com alto consumo E temperatura acima da média: 10


Interpretação e Comparação dos Resultados:

DataFrame 1 (Apenas Consumo > 70%): Encontramos [inserir qtd_alto_consumo] registros que representam [inserir percentual_alto]% da amostra.

DataFrame 2 (Consumo > 70% + Temp T1 > Média): O número de registros caiu para [inserir qtd_consumo_e_temp].

Efeito da inclusão da temperatura: A introdução do segundo critério reduziu o tamanho do DataFrame resultante. Isso indica que nem todos os momentos de pico de consumo de eletrodomésticos ocorrem em condições de temperatura ambiente acima da média. (Complemente com sua própria visão sobre se o uso dos aparelhos parece ter forte correlação com os momentos mais quentes do dia ou não, com base na queda da quantidade de registros).

##Dataset 2 — Steel Industry Energy Consumption (UCI)

#####Fonte: https://archive.ics.uci.edu/dataset/851/steel+industry+energy+consumption

**Situação**

A equipe de gestão de energia de uma indústria siderúrgica deseja localizar situações de consumo elevado e verificar se esses períodos coincidem com condições de carga elevada ou valores menos favoráveis de fator de potência.


In [ ]:
#Realizar a importacao do arquivo
dados = pd.read_csv('Steel_industry_data.csv')
display(dados.head())

,date,Usage_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,CO2(tCO2),Lagging_Current_Power_Factor,Leading_Current_Power_Factor,NSM,WeekStatus,Day_of_week,Load_Type
0,01/01/2018 00:15,3.17,2.95,0.0,0.0,73.21,100.0,900,Weekday,Monday,Light_Load
1,01/01/2018 00:30,4.00,4.46,0.0,0.0,66.77,100.0,1800,Weekday,Monday,Light_Load
2,01/01/2018 00:45,3.24,3.28,0.0,0.0,70.28,100.0,2700,Weekday,Monday,Light_Load
3,01/01/2018 01:00,3.31,3.56,0.0,0.0,68.09,100.0,3600,Weekday,Monday,Light_Load
4,01/01/2018 01:15,3.82,4.50,0.0,0.0,64.72,100.0,4500,Weekday,Monday,Light_Load


In [ ]:
# Selecionar apenas as colunas solicitadas
colunas_selecionadas = [
    'Usage_kWh',
    'Lagging_Current_Reactive.Power_kVarh',
    'Leading_Current_Reactive_Power_kVarh',
    'Lagging_Current_Power_Factor',
    'Leading_Current_Power_Factor',
    'WeekStatus',
    'Day_of_week',
    'Load_Type'
]

dados_selecionados = dados[colunas_selecionadas].copy()

display(dados_selecionados.head())

,Usage_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Lagging_Current_Power_Factor,Leading_Current_Power_Factor,WeekStatus,Day_of_week,Load_Type
0,3.17,2.95,0.0,73.21,100.0,Weekday,Monday,Light_Load
1,4.00,4.46,0.0,66.77,100.0,Weekday,Monday,Light_Load
2,3.24,3.28,0.0,70.28,100.0,Weekday,Monday,Light_Load
3,3.31,3.56,0.0,68.09,100.0,Weekday,Monday,Light_Load
4,3.82,4.50,0.0,64.72,100.0,Weekday,Monday,Light_Load


In [ ]:
# Visualizar os valores de Load_Type e WeekStatus
print("--- Valores de Load_Type ---")
print(dados_selecionados['Load_Type'].value_counts())

print("\n--- Valores de WeekStatus ---")
print(dados_selecionados['WeekStatus'].value_counts())

--- Valores de Load_Type ---
Load_Type
Light_Load      18072
Medium_Load      9696
Maximum_Load     7272
Name: count, dtype: int64

--- Valores de WeekStatus ---
WeekStatus
Weekday    25056
Weekend     9984
Name: count, dtype: int64


In [ ]:
# Verificar valores ausentes
print("--- Valores ausentes por coluna ---")
print(dados_selecionados.isnull().sum())

print("\nTotal de valores ausentes:")
print(dados_selecionados.isnull().sum().sum())

--- Valores ausentes por coluna ---
Usage_kWh                               0
Lagging_Current_Reactive.Power_kVarh    0
Leading_Current_Reactive_Power_kVarh    0
Lagging_Current_Power_Factor            0
Leading_Current_Power_Factor            0
WeekStatus                              0
Day_of_week                             0
Load_Type                               0
dtype: int64

Total de valores ausentes:
0


In [ ]:
# Criar uma amostra aleatória contendo 20% dos registros
amostra = dados_selecionados.sample(
    frac=0.20,
    random_state=42
)

print("Quantidade de registros do dataset original:", len(dados_selecionados))
print("Quantidade de registros da amostra:", len(amostra))

display(amostra.head())

Quantidade de registros do dataset original: 35040
Quantidade de registros da amostra: 7008


,Usage_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Lagging_Current_Power_Factor,Leading_Current_Power_Factor,WeekStatus,Day_of_week,Load_Type
11356,2.88,3.82,0.0,60.20,100.00,Weekend,Sunday,Light_Load
26543,60.77,48.02,0.0,78.46,100.00,Weekday,Thursday,Maximum_Load
2445,120.42,59.65,0.0,89.61,100.00,Weekday,Friday,Maximum_Load
14649,3.13,0.00,16.6,100.00,18.53,Weekend,Saturday,Light_Load
32699,58.86,20.99,0.0,94.19,100.00,Weekday,Friday,Medium_Load


In [ ]:
# Exportar a amostra para CSV
amostra.to_csv(
    'exercicio2steel_sample.csv',
    index=False
)

print("Arquivo exercicio2_sample.csv criado com sucesso.")

Arquivo exercicio2_sample.csv criado com sucesso.


In [ ]:
# Carregar a amostra criada na Etapa A
dados = pd.read_csv('exercicio2steel_sample.csv')

# Renomear as colunas solicitadas
dados = dados.rename(columns={
    'Usage_kWh': 'Consumo_kWh',
    'Lagging_Current_Power_Factor': 'Fator_Potencia_Atrasado',
    'Leading_Current_Power_Factor': 'Fator_Potencia_Adiantado'
})

print(dados.columns)

Index(['Consumo_kWh', 'Lagging_Current_Reactive.Power_kVarh',
       'Leading_Current_Reactive_Power_kVarh', 'Fator_Potencia_Atrasado',
       'Fator_Potencia_Adiantado', 'WeekStatus', 'Day_of_week', 'Load_Type'],
      dtype='object')


In [ ]:
# Inspeção inicial: head, shape, info e describe
print("--- Primeiras linhas ---")
display(dados.head())

print("\n--- Dimensões (Linhas, Colunas) ---")
print(dados.shape)

print("\n--- Estrutura e Tipos de Dados ---")
dados.info()

print("\n--- Estatísticas Descritivas ---")
display(dados.describe())

--- Primeiras linhas ---


,Consumo_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Fator_Potencia_Atrasado,Fator_Potencia_Adiantado,WeekStatus,Day_of_week,Load_Type
0,2.88,3.82,0.0,60.20,100.00,Weekend,Sunday,Light_Load
1,60.77,48.02,0.0,78.46,100.00,Weekday,Thursday,Maximum_Load
2,120.42,59.65,0.0,89.61,100.00,Weekday,Friday,Maximum_Load
3,3.13,0.00,16.6,100.00,18.53,Weekend,Saturday,Light_Load
4,58.86,20.99,0.0,94.19,100.00,Weekday,Friday,Medium_Load



--- Dimensões (Linhas, Colunas) ---
(7008, 8)

--- Estrutura e Tipos de Dados ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7008 entries, 0 to 7007
Data columns (total 8 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Consumo_kWh                           7008 non-null   float64
 1   Lagging_Current_Reactive.Power_kVarh  7008 non-null   float64
 2   Leading_Current_Reactive_Power_kVarh  7008 non-null   float64
 3   Fator_Potencia_Atrasado               7008 non-null   float64
 4   Fator_Potencia_Adiantado              7008 non-null   float64
 5   WeekStatus                            7008 non-null   object 
 6   Day_of_week                           7008 non-null   object 
 7   Load_Type                             7008 non-null   object 
dtypes: float64(5), object(3)
memory usage: 438.1+ KB

--- Estatísticas Descritivas ---


,Consumo_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Fator_Potencia_Atrasado,Fator_Potencia_Adiantado
count,7008.000000,7008.000000,7008.000000,7008.000000,7008.000000
mean,27.790164,13.206839,3.803711,80.672277,84.491448
std,33.717719,16.444246,7.356412,18.869855,30.314798
min,2.450000,0.000000,0.000000,37.870000,12.710000
25%,3.200000,2.380000,0.000000,63.602500,99.710000
50%,4.570000,5.040000,0.000000,87.990000,100.000000
75%,52.107500,22.977500,1.857500,98.842500,100.000000
max,153.140000,87.700000,27.650000,100.000000,100.000000


In [ ]:
# Encontrar o maior consumo registrado
consumo_maximo = dados['Consumo_kWh'].max()

# Calcular 75% do consumo máximo
limiar_consumo = consumo_maximo * 0.75

print(f"Maior consumo registrado: {consumo_maximo:.2f} kWh")
print(f"Limiar de 75% do consumo máximo: {limiar_consumo:.2f} kWh")

Maior consumo registrado: 153.14 kWh
Limiar de 75% do consumo máximo: 114.85 kWh


In [ ]:
# Criar DataFrame com registros acima de 75% do consumo máximo
alto_consumo = dados[
    dados['Consumo_kWh'] > limiar_consumo
].copy()

quantidade_alto_consumo = len(alto_consumo)

percentual_alto_consumo = (
    quantidade_alto_consumo / len(dados)
) * 100

print("Quantidade de registros com consumo elevado:", quantidade_alto_consumo)
print(f"Percentual da amostra: {percentual_alto_consumo:.2f}%")

display(alto_consumo)

Quantidade de registros com consumo elevado: 129
Percentual da amostra: 1.84%


,Consumo_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Fator_Potencia_Atrasado,Fator_Potencia_Adiantado,WeekStatus,Day_of_week,Load_Type
2,120.42,59.65,0.0,89.61,100.0,Weekday,Friday,Maximum_Load
62,118.22,54.04,0.0,90.95,100.0,Weekday,Thursday,Medium_Load
100,129.13,46.58,0.0,94.07,100.0,Weekday,Monday,Medium_Load
108,117.07,46.44,0.0,92.95,100.0,Weekday,Monday,Medium_Load
159,139.57,58.90,0.0,92.13,100.0,Weekday,Thursday,Maximum_Load
...,...,...,...,...,...,...,...,...
6857,124.49,51.73,0.0,92.34,100.0,Weekday,Thursday,Light_Load
6865,117.54,42.91,0.0,93.94,100.0,Weekday,Wednesday,Maximum_Load
6932,123.80,62.24,0.0,89.34,100.0,Weekday,Friday,Maximum_Load
6933,115.81,48.60,0.0,92.21,100.0,Weekday,Wednesday,Medium_Load


In [ ]:
# Normalizar os textos para funcionar tanto com
# "Maximum_Load" quanto com "Maximum Load"
load_type_normalizado = (
    alto_consumo['Load_Type']
    .astype(str)
    .str.replace('_', ' ', regex=False)
    .str.strip()
    .str.lower()
)

quantidade_maximum_load = (
    load_type_normalizado == 'maximum load'
).sum()

print(
    "Registros de alto consumo classificados como Maximum Load:",
    quantidade_maximum_load
)

if quantidade_alto_consumo > 0:
    percentual_maximum_load = (
        quantidade_maximum_load / quantidade_alto_consumo
    ) * 100

    print(
        f"Percentual dos registros de alto consumo que são Maximum Load: "
        f"{percentual_maximum_load:.2f}%"
    )

Registros de alto consumo classificados como Maximum Load: 62
Percentual dos registros de alto consumo que são Maximum Load: 48.06%


In [ ]:
# Vamos analisar o Fator de Potência Atrasado
print("--- Estatísticas do Fator de Potência Atrasado ---")
print(dados['Fator_Potencia_Atrasado'].describe())

print("\n--- Menores valores encontrados ---")
print(
    dados['Fator_Potencia_Atrasado']
    .sort_values()
    .head(20)
    .to_string(index=False)
)

--- Estatísticas do Fator de Potência Atrasado ---
count    7008.000000
mean       80.672277
std        18.869855
min        37.870000
25%        63.602500
50%        87.990000
75%        98.842500
max       100.000000
Name: Fator_Potencia_Atrasado, dtype: float64

--- Menores valores encontrados ---
37.87
39.10
39.27
39.36
40.05
40.12
40.15
40.20
40.62
40.93
41.01
41.10
41.12
41.12
41.19
41.25
41.25
41.48
41.48
41.53


In [ ]:
# Utilizar o primeiro quartil (25%) como limite
# para representar os valores mais baixos de fator de potência
limite_fator_potencia = dados[
    'Fator_Potencia_Atrasado'
].quantile(0.25)

print(
    f"Limite escolhido para fator de potência baixo: "
    f"{limite_fator_potencia:.2f}%"
)

Limite escolhido para fator de potência baixo: 63.60%


In [ ]:
# Criar DataFrame com as duas condições simultaneamente:
# 1. Consumo acima de 75% do máximo
# 2. Fator de potência abaixo do limite escolhido

alto_consumo_fp_baixo = dados[
    (dados['Consumo_kWh'] > limiar_consumo) &
    (dados['Fator_Potencia_Atrasado'] < limite_fator_potencia)
].copy()

quantidade_consumo_fp_baixo = len(alto_consumo_fp_baixo)

percentual_consumo_fp_baixo = (
    quantidade_consumo_fp_baixo / len(dados)
) * 100

print(
    "Quantidade de registros com consumo elevado e fator de potência baixo:",
    quantidade_consumo_fp_baixo
)

print(
    f"Percentual da amostra: "
    f"{percentual_consumo_fp_baixo:.2f}%"
)

display(alto_consumo_fp_baixo)

Quantidade de registros com consumo elevado e fator de potência baixo: 0
Percentual da amostra: 0.00%


,Consumo_kWh,Lagging_Current_Reactive.Power_kVarh,Leading_Current_Reactive_Power_kVarh,Fator_Potencia_Atrasado,Fator_Potencia_Adiantado,WeekStatus,Day_of_week,Load_Type


In [ ]:
print("--- Interpretação e Comparação dos Resultados ---\n")

print(
    f"DataFrame 1 (Consumo > 75% do máximo): "
    f"Foram encontrados {quantidade_alto_consumo} registros, "
    f"representando {percentual_alto_consumo:.2f}% da amostra."
)

print(
    f"\nDesses registros, {quantidade_maximum_load} pertencem "
    f"à categoria Maximum Load."
)

print(
    f"\nFoi utilizado o Fator de Potência Atrasado e definido como "
    f"baixo um valor inferior a {limite_fator_potencia:.2f}%, "
    f"correspondente ao primeiro quartil da amostra."
)

print(
    f"\nDataFrame 2 (Consumo elevado + fator de potência baixo): "
    f"Foram encontrados {quantidade_consumo_fp_baixo} registros, "
    f"representando {percentual_consumo_fp_baixo:.2f}% da amostra."
)

print(
    "\nEsses registros merecem maior atenção da equipe de energia porque "
    "combinam um nível elevado de consumo com valores relativamente baixos "
    "de fator de potência. Essa combinação pode indicar momentos em que a "
    "energia elétrica está sendo utilizada de forma menos eficiente, com "
    "maior participação de potência reativa. Portanto, esses períodos são "
    "bons candidatos para investigação e possíveis ações de melhoria na "
    "eficiência energética da indústria."
)

--- Interpretação e Comparação dos Resultados ---

DataFrame 1 (Consumo > 75% do máximo): Foram encontrados 129 registros, representando 1.84% da amostra.

Desses registros, 62 pertencem à categoria Maximum Load.

Foi utilizado o Fator de Potência Atrasado e definido como baixo um valor inferior a 63.60%, correspondente ao primeiro quartil da amostra.

DataFrame 2 (Consumo elevado + fator de potência baixo): Foram encontrados 0 registros, representando 0.00% da amostra.

Esses registros merecem maior atenção da equipe de energia porque combinam um nível elevado de consumo com valores relativamente baixos de fator de potência. Essa combinação pode indicar momentos em que a energia elétrica está sendo utilizada de forma menos eficiente, com maior participação de potência reativa. Portanto, esses períodos são bons candidatos para investigação e possíveis ações de melhoria na eficiência energética da indústria.


## Dataset 3 — Power Consumption of Tetouan City (UCI)


#####Fonte: https://archive.ics.uci.edu/dataset/849/power+consumption+of+tetouan+city

**Situação**

Uma equipe responsável pelo planejamento da distribuição de energia em Tétouan deseja identificar qual das três zonas apresenta o maior pico de consumo e observar as condições ambientais registradas nos momentos de maior demanda.


In [ ]:
# Carregar conjunto de dados
df = pd.read_csv('exercicio3_sample.csv')

In [ ]:
# Remova metadados/tipo linhas geradas pelo Orange Data Mining, se presente
df = df[pd.to_numeric(df['Temperature'], errors='coerce').notna()].copy()

In [ ]:
# Converter colunas numéricas em float
for col in df.columns:
    df[col] = pd.to_numeric(df[col])

In [ ]:
# 1. Renomeação das colunas
df.rename(
    columns={
        'PowerConsumption_Zone1': 'Consumo_Zona_1',
        'PowerConsumption_Zone2': 'Consumo_Zona_2',
        'PowerConsumption_Zone3': 'Consumo_Zona_3',
    },
    inplace=True,
)

print('=== 1. Renomeação de Colunas ===')
print('Colunas renomeadas com sucesso para:')
print(' - Consumo_Zona_1\n - Consumo_Zona_2\n - Consumo_Zona_3\n')

=== 1. Renomeação de Colunas ===
Colunas renomeadas com sucesso para:
 - Consumo_Zona_1
 - Consumo_Zona_2
 - Consumo_Zona_3



In [ ]:
# 2. Consumo Máximo por Zona
max_z1 = df['Consumo_Zona_1'].max()
max_z2 = df['Consumo_Zona_2'].max()
max_z3 = df['Consumo_Zona_3'].max()

print('=== 2. Consumo Máximo por Zona ===')
print(f'Consumo_Zona_1: {max_z1:,.2f} kW')
print(f'Consumo_Zona_2: {max_z2:,.2f} kW')
print(f'Consumo_Zona_3: {max_z3:,.2f} kW\n')

=== 2. Consumo Máximo por Zona ===
Consumo_Zona_1: 51,571.50 kW
Consumo_Zona_2: 36,437.20 kW
Consumo_Zona_3: 47,441.70 kW



In [ ]:
# 3. Maior Pico de Consumo
print('=== 3. Maior Pico de Consumo ===')
print(
    f'Maior Pico Global: Consumo_Zona_1 com o pico de {max_z1:,.2f} kW.\n'
)

=== 3. Maior Pico de Consumo ===
Maior Pico Global: Consumo_Zona_1 com o pico de 51,571.50 kW.



In [ ]:
# 4. Limiar de 70%
limiar_70 = 0.70 * max_z1

print('=== 4. Cálculo do Limiar de 70% ===')
print(f'Calculo: 0,70 * {max_z1:,.2f} kW')
print(f'Limiar de 70%: {limiar_70:,.2f} kW\n')

=== 4. Cálculo do Limiar de 70% ===
Calculo: 0,70 * 51,571.50 kW
Limiar de 70%: 36,100.05 kW



In [ ]:
# 5. Contagem e Percentual (DataFrame 1)
df1 = df[df['Consumo_Zona_1'] > limiar_70]
total_amostra = len(df)
qtd_df1 = len(df1)
pct_df1 = (qtd_df1 / total_amostra) * 100

print('=== 5. Contagem e Percentual (DataFrame 1) ===')
print(f'Total de registros na amostra: {total_amostra}')
print(f'Registros acima do limiar (df1): {qtd_df1}')
print(f'Percentual da amostra total: {pct_df1:.2f}%\n')

=== 5. Contagem e Percentual (DataFrame 1) ===
Total de registros na amostra: 7863
Registros acima do limiar (df1): 2388
Percentual da amostra total: 30.37%



In [ ]:
# 6. Temperatura Média e DataFrame 2
temp_media = df['Temperature'].mean()
df2 = df1[df1['Temperature'] > temp_media]
qtd_df2 = len(df2)
pct_df2 = (qtd_df2 / total_amostra) * 100

print('=== 6. Temperatura Média e DataFrame 2 ===')
print(f'Temperatura Média Geral: {temp_media:.2f} °C')
print(
    f'Registros filtrados em df2 (Consumo > {limiar_70:,.2f} kW E Temp > {temp_media:.2f} °C): {qtd_df2}'
)
print(f'Percentual da amostra total: {pct_df2:.2f}%\n')

=== 6. Temperatura Média e DataFrame 2 ===
Temperatura Média Geral: 18.79 °C
Registros filtrados em df2 (Consumo > 36,100.05 kW E Temp > 18.79 °C): 1597
Percentual da amostra total: 20.31%



In [ ]:
# 7. Comparação e Explicação
reducao = qtd_df1 - qtd_df2
pct_reducao = (reducao / qtd_df1) * 100
pct_coincide = (qtd_df2 / qtd_df1) * 100

print('=== 7. Comparação e Explicação ===')
print(
    f'Redução: A inclusão do filtro de temperatura reduziu o conjunto de {qtd_df1} para {qtd_df2} registros.'
)
print(
    f'Diferença: Queda de {reducao} registros (~{pct_reducao:.1f}% dos registros de pico).\n'
)
print('Análise:')
print(
    f' - {pct_coincide:.1f}% dos picos da Zona 1 coincidem com temperaturas acima da média ({temp_media:.2f} °C),'
)
print(
    '   o que aponta padrão típico do uso de equipamentos de refrigeração e ar-condicionado.'
)
print(
    f' - {pct_reducao:.1f}% dos picos ocorrem sob temperaturas amenas/baixas,'
)
print(
    '   sendo induzidos por outros fatores de demanda industrial, comercial ou operacional.'
)

=== 7. Comparação e Explicação ===
Redução: A inclusão do filtro de temperatura reduziu o conjunto de 2388 para 1597 registros.
Diferença: Queda de 791 registros (~33.1% dos registros de pico).

Análise:
 - 66.9% dos picos da Zona 1 coincidem com temperaturas acima da média (18.79 °C),
   o que aponta padrão típico do uso de equipamentos de refrigeração e ar-condicionado.
 - 33.1% dos picos ocorrem sob temperaturas amenas/baixas,
   sendo induzidos por outros fatores de demanda industrial, comercial ou operacional.


## Dataset 4 — Solar Power Generation Data (Kaggle)

#####Fonte: https://www.kaggle.com/datasets/anikannal/solar-power-generation-data

**Situação**

Uma equipe de operação de uma usina fotovoltaica deseja localizar períodos de alta geração e identificar quais inversores aparecem com maior frequência nesses momentos. Nesta atividade, o foco inicial será o arquivo de geração da planta, sem realizar junção com o arquivo de sensores.


In [ ]:
dados = pd.read_csv('Plant_1_Generation_Data.csv')

print("--- Primeiras linhas ---")
display(dados.head())
print("\n--- Dimensões (Linhas, Colunas) ---")
print(dados.shape)
print("\n--- Estrutura e Tipos de Dados ---")
dados.info()
print("\n--- Estatísticas Descritivas ---")
display(dados.describe())

--- Primeiras linhas ---


,DATE_TIME,PLANT_ID,SOURCE_KEY,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
0,15-05-2020 00:00,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0
1,15-05-2020 00:00,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0
2,15-05-2020 00:00,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0
3,15-05-2020 00:00,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0
4,15-05-2020 00:00,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0



--- Dimensões (Linhas, Colunas) ---
(68778, 7)

--- Estrutura e Tipos de Dados ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 68778 entries, 0 to 68777
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   DATE_TIME    68778 non-null  object 
 1   PLANT_ID     68778 non-null  int64  
 2   SOURCE_KEY   68778 non-null  object 
 3   DC_POWER     68778 non-null  float64
 4   AC_POWER     68778 non-null  float64
 5   DAILY_YIELD  68778 non-null  float64
 6   TOTAL_YIELD  68778 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 3.7+ MB

--- Estatísticas Descritivas ---


,PLANT_ID,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
count,68778.0,68778.000000,68778.000000,68778.000000,6.877800e+04
mean,4135001.0,3147.426211,307.802752,3295.968737,6.978712e+06
std,0.0,4036.457169,394.396439,3145.178309,4.162720e+05
min,4135001.0,0.000000,0.000000,0.000000,6.183645e+06
25%,4135001.0,0.000000,0.000000,0.000000,6.512003e+06
50%,4135001.0,429.000000,41.493750,2658.714286,7.146685e+06
75%,4135001.0,6366.964286,623.618750,6274.000000,7.268706e+06
max,4135001.0,14471.125000,1410.950000,9163.000000,7.846821e+06


In [ ]:
dados = dados.rename(columns={
    "DATE_TIME": "Data_Hora",
    "PLANT_ID": "ID_Planta",
    "SOURCE_KEY": "Inversor",
    "DC_POWER": "Potencia_CC",
    "AC_POWER": "Potencia_CA",
    "DAILY_YIELD": "Producao_Diaria",
    "TOTAL_YIELD": "Producao_Total"
})

print(dados.columns)

Index(['Data_Hora', 'ID_Planta', 'Inversor', 'Potencia_CC', 'Potencia_CA',
       'Producao_Diaria', 'Producao_Total'],
      dtype='object')


In [ ]:
geracao_maxima = dados['Potencia_CA'].max()
print(f"Maior geração registrada: {geracao_maxima} kW")

limiar_70 = 0.70 * geracao_maxima
print(f"Limiar de 70% do máximo: {limiar_70} kW")
df_alta_geracao = dados[dados['Potencia_CA'] > limiar_70]

Maior geração registrada: 1410.95 kW
Limiar de 70% do máximo: 987.665 kW


In [ ]:
qtd_alta_geracao = len(df_alta_geracao)
percentual_alto = (qtd_alta_geracao / len(dados)) * 100

print(f"Quantidade de registros acima de 70%: {qtd_alta_geracao}")
print(f"Percentual que representam da amostra: {percentual_alto:.2f}%")

Quantidade de registros acima de 70%: 5996
Percentual que representam da amostra: 8.72%


In [ ]:
frequencia_inversores = df_alta_geracao['Inversor'].value_counts()
print("Frequência de cada inversor nos registros de alta geração:")
display(frequencia_inversores)

inversor_mais_frequente = frequencia_inversores.idxmax()
qtd_inversor_mais_frequente = frequencia_inversores.max()

print(f"Inversor mais frequente nos picos de geração: {inversor_mais_frequente}")
print(f"Número de vezes que ele aparece: {qtd_inversor_mais_frequente}")

Frequência de cada inversor nos registros de alta geração:


,count
Inversor,
adLQvlD726eNBSB,310
McdE0feGgRqW7Ca,301
ZnxXDlPa8U1GXgE,298
1IF53ai7Xc0U56Y,296
VHMLBKoKgIrUVDU,291
iCRJl6heRkivqQ3,289
zVJPv84UY57bAof,289
wCURE6d3bPkepu2,288
3PZuoBAID5Wc2HD,285


Inversor mais frequente nos picos de geração: adLQvlD726eNBSB
Número de vezes que ele aparece: 310


Interpretação e Comparação dos Resultados

Depois de aplicar o filtro de 70% do valor máximo de Potência_CA, obtive 5.996 registros, o que dá cerca de 8,72% do total da amostra. Achei esse número baixo mas fez sentido pensando no contexto: como é geração solar, só faz sentido ter valores altos perto do meio-dia, quando o sol está mais forte. No resto do dia (de manhã cedo, fim de tarde, à noite) a geração cai bastante ou some, então é normal que só uma fatia pequena dos dados fique acima desse limiar.

Depois, olhando pra frequência de cada inversor dentro desses registros de pico, o que mais aparece é o adLQvlD726eNBSB, com 310 vezes. Mas quando olhei a lista toda dos 22 inversores, os valores ficam bem próximos uns dos outros (a maioria entre 250 e 310), então não dá pra dizer que só um inversor está "carregando" a geração da planta.

O que chamou atenção foram dois inversores com número bem mais baixo que os outros: o 1BY6WEcLGh8j5v7 (201) e o bvBOhCH3iADSZry (184). Como esses dois ficaram destacados abaixo da média dos demais, acho que pode ser algo específico daquele ponto da instalação, tipo sombra de alguma construção/árvore perto, painel sujo ou até algum problema técnico que faz ele gerar menos nos horários de pico. Seria interessante a equipe de manutenção dar uma olhada nesses dois inversores especificamente pra entender se é algo pontual ou recorrente.

## Dataset 5 — Wind & Solar Energy Production (Kaggle)


#####Fonte: https://www.kaggle.com/datasets/ahmeduzaki/wind-and-solar-energy-production-dataset

In [ ]:
file_path = "Energy Production Dataset.csv"
df = pd.read_csv(file_path)

print("=== ETAPA A.1: Atributos disponíveis ===")
print("Colunas:", df.columns.tolist())
print("\nPrimeiras 5 linhas:")
print(df.head())

colunas_temporais = [
    "Date",
    "Start_Hour",
    "End_Hour",
    "Day_of_Year",
    "Day_Name",
    "Month_Name",
    "Season",
]
colunas_selecionadas = colunas_temporais + ["Source", "Production"]
df_selecionado = df[colunas_selecionadas].copy()

print("\n=== ETAPA A.2: Atributos Mantidos ===")
print(df_selecionado.head())

solar_min = df_selecionado[df_selecionado["Source"] == "Solar"][
    "Production"
].min()
solar_max = df_selecionado[df_selecionado["Source"] == "Solar"][
    "Production"
].max()

eolica_min = df_selecionado[df_selecionado["Source"] == "Wind"][
    "Production"
].min()
eolica_max = df_selecionado[df_selecionado["Source"] == "Wind"][
    "Production"
].max()

print("\n=== ETAPA A.3: Valores Mínimos e Máximos ===")
print(f"Geração Solar  -> Mínimo: {solar_min} | Máximo: {solar_max}")
print(f"Geração Eólica -> Mínimo: {eolica_min} | Máximo: {eolica_max}")

print("\n=== ETAPA A.4: Valores Ausentes ===")
print(df_selecionado.isnull().sum())

df_amostra = df_selecionado.sample(frac=0.20, random_state=42)

df_amostra.to_csv("amostra_20_percent.csv", index=False)
print(
    f"\n=== ETAPA A.5 & A.6: Amostra gerada com {len(df_amostra)} registros e salva em 'amostra_20_percent.csv' ==="
)

df_solar = (
    df_selecionado[df_selecionado["Source"] == "Solar"]
    .copy()
    .rename(columns={"Production": "Geracao_Solar"})
)
df_eolica = (
    df_selecionado[df_selecionado["Source"] == "Wind"]
    .copy()
    .rename(columns={"Production": "Geracao_Eolica"})
)

max_solar = df_solar["Geracao_Solar"].max()
max_eolica = df_eolica["Geracao_Eolica"].max()

print("\n=== ETAPA B.2: Valores Máximos Registrados ===")
print(f"Máximo Solar : {max_solar} MW")
print(f"Máximo Eólico: {max_eolica} MW")

limite_solar_70 = max_solar * 0.70
limite_eolico_70 = max_eolica * 0.70

print("\n=== ETAPA B.3: Limites de 70% do Máximo ===")
print(f"Limite 70% Solar : {limite_solar_70:.2f} MW")
print(f"Limite 70% Eólico: {limite_eolico_70:.2f} MW")

df_alta_solar = df_solar[df_solar["Geracao_Solar"] >= limite_solar_70]
df_alta_eolica = df_eolica[df_eolica["Geracao_Eolica"] >= limite_eolico_70]

total_registros_solar = len(df_solar)
total_registros_eolica = len(df_eolica)

qtd_alta_solar = len(df_alta_solar)
qtd_alta_eolica = len(df_alta_eolica)

pct_alta_solar = (qtd_alta_solar / total_registros_solar) * 100
pct_alta_eolica = (qtd_alta_eolica / total_registros_eolica) * 100

print("\n=== ETAPA B.5: Frequência e Percentuais de Alta Geração ===")
print(
    f"Alta Geração Solar  (>= 70%): {qtd_alta_solar} registros de {total_registros_solar} ({pct_alta_solar:.2f}%)"
)
print(
    f"Alta Geração Eólica (>= 70%): {qtd_alta_eolica} registros de {total_registros_eolica} ({pct_alta_eolica:.2f}%)"
)

print("\n=== ETAPA B.6: Comparação de Frequência ===")
if pct_alta_solar > pct_alta_eolica:
    print(
        f"A fonte SOLAR apresentou maior frequência proporcional acima de 70% do seu topo ({pct_alta_solar:.2f}% vs {pct_alta_eolica:.2f}%)."
    )
elif pct_alta_eolica > pct_alta_solar:
    print(
        f"A fonte EÓLICA apresentou maior frequência proporcional acima de 70% do seu topo ({pct_alta_eolica:.2f}% vs {pct_alta_solar:.2f}%)."
    )
else:
    print("Ambas as fontes possuem exatamente o mesmo percentual relativo.")

=== ETAPA A.1: Atributos disponíveis ===
Colunas: ['Date', 'Start_Hour', 'End_Hour', 'Source', 'Day_of_Year', 'Day_Name', 'Month_Name', 'Season', 'Production']

Primeiras 5 linhas:
         Date  Start_Hour  End_Hour Source  Day_of_Year Day_Name Month_Name  \
0  11/30/2025          21        22   Wind          334   Sunday   November   
1  11/30/2025          18        19   Wind          334   Sunday   November   
2  11/30/2025          16        17   Wind          334   Sunday   November   
3  11/30/2025          23         0   Wind          334   Sunday   November   
4  11/30/2025           6         7   Wind          334   Sunday   November   

  Season  Production  
0   Fall        5281  
1   Fall        3824  
2   Fall        3824  
3   Fall        6120  
4   Fall        4387  

=== ETAPA A.2: Atributos Mantidos ===
         Date  Start_Hour  End_Hour  Day_of_Year Day_Name Month_Name Season  \
0  11/30/2025          21        22          334   Sunday   November   Fall   
1  11/30/

**Situação**

Um operador de um portfólio de fontes renováveis deseja comparar a ocorrência de períodos de alta produção solar e alta produção eólica. Como as duas fontes possuem escalas próprias, cada uma deverá ser comparada com o seu próprio valor máximo.


## Dataset 6 — Individual Household Electric Power Consumption (UCI)

#####Fonte: https://archive.ics.uci.edu/ml/datasets/individual+household+electric+power+consumption

Etapa A B

In [ ]:
import io
import zipfile
import pandas as pd
from google.colab import files

print("=== PIPELINE COMPLETO (ETAPAS A e B) ===")

# ---------------------------------------------------------
# ETAPA A: UPLOAD, SELEÇÃO, TRATAMENTO E AMOSTRAGEM
# ---------------------------------------------------------
print("\n1. Faça o upload do arquivo bruto (.csv, .txt ou .zip):")
uploaded = files.upload()
file_name = list(uploaded.keys())[0]

# Tratamento para extrair arquivo caso seja enviado em formato .ZIP
if file_name.endswith(".zip"):
  with zipfile.ZipFile(io.BytesIO(uploaded[file_name]), "r") as z:
    # Pega o primeiro arquivo dentro do zip que seja csv ou txt
    internal_files = [
        f for f in z.namelist() if f.endswith(".csv") or f.endswith(".txt")
    ]
    extracted_file = z.open(internal_files[0])
    # Tenta ler com UTF-8 e fallback para ISO-8859-1 se houver caracteres especiais
    try:
      df_raw = pd.read_csv(extracted_file, sep=None, engine="python")
    except UnicodeDecodeError:
      extracted_file.seek(0)
      df_raw = pd.read_csv(
          extracted_file, sep=None, engine="python", encoding="iso-8859-1"
      )
else:
  # Carregamento direto caso não seja zip
  try:
    df_raw = pd.read_csv(
        io.BytesIO(uploaded[file_name]),
        sep=";",
        na_values="?",
        low_memory=False,
    )
  except Exception:
    df_raw = pd.read_csv(
        io.BytesIO(uploaded[file_name]), sep=",", low_memory=False
    )

# Mapeamento dinâmico de colunas (Atende tanto Household Power quanto Appliances Energy)
colunas_household = [
    "Global_active_power",
    "Global_reactive_power",
    "Voltage",
    "Global_intensity",
    "Sub_metering_1",
    "Sub_metering_2",
    "Sub_metering_3",
]
colunas_appliances = [
    "Appliances",
    "lights",
    "T1",
    "RH_1",
    "T2",
    "RH_2",
    "Press_mm_hg",
]

if all(col in df_raw.columns for col in colunas_household):
  colunas_desejadas = colunas_household
elif all(col in df_raw.columns for col in colunas_appliances):
  colunas_desejadas = colunas_appliances
else:
  # Caso as colunas tenham nomes ligeiramente diferentes, seleciona as 7 primeiras numéricas
  colunas_desejadas = df_raw.select_dtypes(include=["number"]).columns[:7].tolist()

df = df_raw[colunas_desejadas].copy()

# Imputação de nulos usando a mediana de cada coluna
for col in df.columns:
  df[col] = pd.to_numeric(df[col], errors="coerce")
  df[col] = df[col].fillna(df[col].median())

# Amostra aleatória de 10%
df_amostra = df.sample(frac=0.10, random_state=42).copy()

# Salvando a amostra tratada em CSV
df_amostra.to_csv("amostra_household_10pct.csv", index=False)

# ---------------------------------------------------------
# ETAPA B: RENOMEAÇÃO, CÁLCULOS E FILTRAGENS
# ---------------------------------------------------------
# Normalizando os nomes para as 7 colunas do exercício
df_amostra.columns = [
    "pot_ativa",
    "pot_reativa",
    "tensao",
    "corrente",
    "sub_1",
    "sub_2",
    "sub_3",
]

# Determinação da Potência Máxima e do Limite de 75%
max_pot_ativa = df_amostra["pot_ativa"].max()
limite_75_pot = max_pot_ativa * 0.75

# DataFrame 1: Potência Ativa > 75% do máximo
df_demanda_alta = df_amostra[df_amostra["pot_ativa"] > limite_75_pot]
qtd_demanda_alta = len(df_demanda_alta)
pct_demanda_alta = (qtd_demanda_alta / len(df_amostra)) * 100

# Cálculo da Corrente Média da amostra
corrente_media = df_amostra["corrente"].mean()

# DataFrame 2: Potência Ativa > 75% E Corrente > Média
df_critico = df_amostra[
    (df_amostra["pot_ativa"] > limite_75_pot)
    & (df_amostra["corrente"] > corrente_media)
]
qtd_critico = len(df_critico)
pct_critico = (qtd_critico / len(df_amostra)) * 100

# ---------------------------------------------------------
# RELATÓRIO FINAL DE SAÍDA
# ---------------------------------------------------------
print("\n" + "=" * 45)
print("             RELATÓRIO DE RESULTADOS         ")
print("=" * 45)
print(f"Total de registros originais:    {len(df_raw)}")
print(f"Total na amostra de 10%:         {len(df_amostra)}")
print(f"Potência Ativa Máxima (Amostra): {max_pot_ativa:.3f}")
print(f"Limite de Corte (75% do Máx):    {limite_75_pot:.3f}")
print(f"Corrente Média da Amostra:       {corrente_media:.2f}\n")

print(
    f"1. Demanda Alta (Pot. Ativa > 75%):"
    f"\n   -> {qtd_demanda_alta} registros ({pct_demanda_alta:.2f}% da amostra)\n"
)

print(
    f"2. Demanda Crítica (Pot. Ativa > 75% E Corrente > Média):"
    f"\n   -> {qtd_critico} registros ({pct_critico:.2f}% da amostra)"
)
print("=" * 45)

=== PIPELINE COMPLETO (ETAPAS A e B) ===

1. Faça o upload do arquivo bruto (.csv, .txt ou .zip):


Saving appliances+energy+prediction.zip to appliances+energy+prediction (1).zip

             RELATÓRIO DE RESULTADOS         
Total de registros originais:    19735
Total na amostra de 10%:         1974
Potência Ativa Máxima (Amostra): 770.000
Limite de Corte (75% do Máx):    577.500
Corrente Média da Amostra:       40.22

1. Demanda Alta (Pot. Ativa > 75%):
   -> 17 registros (0.86% da amostra)

2. Demanda Crítica (Pot. Ativa > 75% E Corrente > Média):
   -> 10 registros (0.51% da amostra)


###Dataset 7 — Compare os dois conjuntos e explique o efeito da inclusão da corrente como segunda condição.

**Comparação dos Conjuntos:**

 **Conjunto 1** (Apenas Potência Ativa > 75% do máximo): Isola unicamente os momentos de consumo elétrico no topo da escala da residência (picos de potência ativa em kW).

**Conjunto 2** (Potência Ativa > 75% do máximo E Corrente > média): Aplica um filtro duplo, exigindo que a alta potência venha acompanhada de uma intensidade de corrente (I, em Amperes) acima do consumo médio da amostra.Efeito da Inclusão da Corrente como Segunda CondiçãoRefinamento do Filtro: A inclusão da corrente funciona como uma validação de estresse no circuito elétrico. A relação básica de potência em sistemas elétricos é aproximada por **Potência =~ Tensão x Corrente**.

Eliminação de Oscilações de Tensão: Momentos em que a potência atinge valores elevados podem ser causados tanto por alta corrente quanto por variações/picos na tensão fornecida pela rede. Exigir que a corrente também seja superior à média garante que o evento corresponda a uma demanda física real de carga (vários aparelhos ou equipamentos de alta potência ligados simultaneamente).
Confirmação de Carga Crítica: Na prática, esse filtro combinado reduz o ruído de medições isoladas e confirma episódios em que a fiação e os disjuntores da residência estão sendo submetidos a uma operação de alta carga contínua.

**Situação**

Uma residência possui monitoramento elétrico detalhado e deseja identificar episódios de demanda elevada que também apresentem corrente acima do comportamento médio. O dataset é o mesmo utilizado como referência em aula, mas o critério de análise será diferente.
